### PR TITLE
DownloadExercise now creates a seperate folder for course

### PR DESCRIPTION
**Pull request by JaniL**
Course folder is named by course's name.

Also Course object now also knows all exercises of it.

### CODE REVIEW

**Comment by JaniL** [comment]
This was part of exercise download user story. We noticed that we missed something in it, so here's the missing part of it.

File: src/main/java/hy/tmc/cli/backendcommunication/ExerciseDownloader.java [hunk 1]
```diff
@@ -58,17 +58,40 @@ public void downloadFiles(List<Exercise> exercises) {
 
     /**
      * Method for downloading files if path where to download is defined.
+     * @param exercises
+     * @param path
```
Comment (jamox) [inline src/main/java/hy/tmc/cli/backendcommunication/ExerciseDownloader.java]: No empty tags, just remove these

File: src/main/java/hy/tmc/cli/backendcommunication/ExerciseDownloader.java [hunk 2]
```diff
@@ -58,17 +58,40 @@ public void downloadFiles(List<Exercise> exercises) {
...
      * @param exercises list of exercises which will be downloaded, list is parsed from json.
      * @param path server path to exercises.
+     * @param folderName folder name of where exercises will be extracted (for example course name)
      */
-    public void downloadFiles(List<Exercise> exercises, String path) {
+    public void downloadFiles(List<Exercise> exercises, String path, String folderName) {
         int exCount = 0;
         path = getCorrectPath(path);
+
+        if (folderName != null && !folderName.isEmpty()) {
```
Comment (jamox) [inline src/main/java/hy/tmc/cli/backendcommunication/ExerciseDownloader.java]: Just use http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Strings.html#isNullOrEmpty(java.lang.String)

File: src/main/java/hy/tmc/cli/backendcommunication/ExerciseDownloader.java [hunk 3]
```diff
@@ -58,17 +58,40 @@ public void downloadFiles(List<Exercise> exercises) {
...
+            coursePath.mkdirs();
+        }
+
         for (Exercise exercise : exercises) {
             handleSingleExercise(exercise, exCount, exercises, path);
             exCount++;
         }
+
         if (this.front != null) {
             front.printLine(exercises.size() + " exercises downloaded.");
```
Comment (jamox) [inline src/main/java/hy/tmc/cli/backendcommunication/ExerciseDownloader.java]: Maybe take the effort to have 1 exercise downloaded and N exercises downloaded. (pluralize correctly)

File: src/main/java/hy/tmc/cli/backendcommunication/TmcJsonParser.java [hunk 4]
```diff
@@ -18,11 +18,7 @@
...
 
     /**
-<<<<<<< HEAD:src/main/java/hy/tmc/cli/backendcommunication/TmcJsonParser.java
      * Get JSON-data from url.
-=======
-     * get JSON-data from url.
->>>>>>> origin:src/main/java/hy/tmc/cli/backendcommunication/TmcJsonParser.java
      *
      * @param url url from which the object data is fetched
      * @return JSON-object containing JSON-data
```
Comment (jamox) [inline src/main/java/hy/tmc/cli/backendcommunication/TmcJsonParser.java]: JSON object contains json data - yeah not really worth saying that.

File: src/main/java/hy/tmc/cli/backendcommunication/TmcJsonParser.java [hunk 5]
```diff
@@ -81,6 +77,28 @@ public static String getExerciseNames(String courseUrl) {
...
+        ConfigHandler confighandler = new ConfigHandler();
+        return getCourse(confighandler.getCourseUrl(courseID));
+    }
+
+    /**
+     * Get information about course specified by the URL path
+     * to course json.
+     * @param courseURL URL path to course json
+     * @return an Course object (parsed from JSON)
+     */
```
Comment (jamox) [inline src/main/java/hy/tmc/cli/backendcommunication/TmcJsonParser.java]: in general JSON (or json) - keep the same case all the time.

File: src/main/java/hy/tmc/cli/backendcommunication/TmcJsonParser.java [hunk 6]
```diff
@@ -81,6 +77,28 @@ public static String getExerciseNames(String courseUrl) {
...
+    public static Course getCourse(String courseURL) {
```
Comment (jamox) [inline src/main/java/hy/tmc/cli/backendcommunication/TmcJsonParser.java]: Variable should be camel case -> `courseUrl` not `courseURL`
see: https://google-styleguide.googlecode.com/svn/trunk/javaguide.html#s5.3-camel-case

File: src/test/java/feature/listcourses/ListCoursesSteps.java [hunk 7]
```diff
@@ -40,10 +45,10 @@
     @Before
     public void setUpServer() throws IOException {
         configHandler = new ConfigHandler();
-        configHandler.writeServerAddress("http://127.0.0.1:8080");
+        configHandler.writeServerAddress("http://127.0.0.1:7777");
```
Comment (jamox) [inline src/test/java/feature/listcourses/ListCoursesSteps.java]: Magic string; why not keep this/these in variables->
```java
private static final String SERVER_URI = "http://127.0.0.1";
private static final String SERVER_PORT = "7777";
private static final String SERVER_ADDRESS = SERVER_URI + ":" + SERVER_PORT;
```

File: src/test/java/feature/login/LoginSteps.java [hunk 8]
```diff
@@ -32,7 +35,7 @@
     @Before
     public void initializeServer() throws IOException {
         configHandler = new ConfigHandler();
-        configHandler.writeServerAddress("http://127.0.0.1:8080");
+        configHandler.writeServerAddress("http://127.0.0.1:3333");
```
Comment (jamox) [inline src/test/java/feature/login/LoginSteps.java]: Same as the previous file.

File: src/test/java/hy/tmc/cli/domain/CourseTest.java [hunk 9]
```diff
@@ -18,6 +22,8 @@
     private Course course;
     private final int id = 7;
     private final String name = "ankka";
+    private List<Exercise> exercises;
+    private Exercise ex;
     
     public CourseTest() {
     }
```
Comment (jamox) [inline src/test/java/hy/tmc/cli/domain/CourseTest.java]: why empty constructor?

**Comment by JaniL** [comment]
Waiting for final feedback from team members before merge

**Pull request merged**